File: Scripts/generate_icon.py
#!/usr/bin/env python3
"""
StatBar App Icon Generator
生成现代化简洁的应用图标
"""

from PIL import Image, ImageDraw
import math
import os

# 图标尺寸
SIZES = [16, 32, 64, 128, 256, 512, 1024]

def create_rounded_rect_mask(size, radius):
    """创建圆角矩形蒙版"""
    mask = Image.new('L', (size, size), 0)
    draw = ImageDraw.Draw(mask)
    draw.rounded_rectangle([(0, 0), (size-1, size-1)], radius=radius, fill=255)
    return mask

def draw_gradient_rect(draw, width, height, color1, color2):
    """绘制渐变矩形"""
    for y in range(height):
        ratio = y / height
        r = int(color1[0] + (color2[0] - color1[0]) * ratio)
        g = int(color1[1] + (color2[1] - color1[1]) * ratio)
        b = int(color1[2] + (color2[2] - color1[2]) * ratio)
        draw.line([(0, y), (width-1, y)], fill=(r, g, b))

def create_icon(size):
    """创建指定尺寸的图标"""
    img = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)
    
    # 计算比例
    scale = size / 1024
    radius = int(180 * scale)
    
    # 绘制圆角背景（渐变：深蓝到紫色）
    for y in range(size):
        ratio = y / size
        # 对角线渐变
        r = int(51 + (128 - 51) * ratio)   # 0.2 -> 0.5
        g = int(102 + (77 - 102) * ratio)  # 0.4 -> 0.3
        b = int(230 + (230 - 230) * ratio) # 0.9 -> 0.9
        draw.line([(0, y), (size-1, y)], fill=(r, g, b, 255))
    
    # 应用圆角蒙版
    mask = create_rounded_rect_mask(size, radius)
    img.putalpha(mask)
    
    # 创建临时图像用于绘制白色元素
    overlay = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    overlay_draw = ImageDraw.Draw(overlay)
    
    # CPU 主体位置
    cpu_x = int(312 * scale)
    cpu_y = int(312 * scale)
    cpu_w = int(400 * scale)
    cpu_h = int(400 * scale)
    cpu_r = int(60 * scale)
    
    # 绘制 CPU 主体（白色圆角矩形）
    overlay_draw.rounded_rectangle(
        [(cpu_x, cpu_y), (cpu_x + cpu_w, cpu_y + cpu_h)],
        radius=cpu_r,
        fill=(255, 255, 255, 242)  # 95% 白色
    )
    
    # 绘制 CPU 引脚
    pin_w = int(40 * scale)
    pin_h = int(100 * scale)
    pin_spacing = int(100 * scale)
    pin_start_x = int(372 * scale)
    
    for i in range(3):
        px = pin_start_x + i * pin_spacing
        
        # 上方引脚
        overlay_draw.rectangle(
            [(px, int(212 * scale)), (px + pin_w, int(212 * scale) + pin_h)],
            fill=(255, 255, 255, 242)
        )
        
        # 下方引脚
        overlay_draw.rectangle(
            [(px, int(712 * scale)), (px + pin_w, int(712 * scale) + pin_h)],
            fill=(255, 255, 255, 242)
        )
        
        # 左侧引脚（旋转90度）
        center_x = size // 2
        center_y = size // 2
        # 简化：直接画上下旋转后的位置
        overlay_draw.rectangle(
            [(int(212 * scale), px), (int(212 * scale) + pin_h, px + pin_w)],
            fill=(255, 255, 255, 242)
        )
        
        # 右侧引脚
        overlay_draw.rectangle(
            [(int(712 * scale), px), (int(712 * scale) + pin_h, px + pin_w)],
            fill=(255, 255, 255, 242)
        )
    
    # 合并图层
    img = Image.alpha_composite(img, overlay)
    draw = ImageDraw.Draw(img)
    
    # CPU 内部图案（三个竖条表示活动状态）- 蓝色
    bar_w = int(60 * scale)
    bar_spacing = int(90 * scale)
    bar_heights = [140, 200, 120]
    bar_start_x = int(412 * scale)
    bar_start_y = int(412 * scale)
    
    for i, h in enumerate(bar_heights):
        bx = bar_start_x + i * bar_spacing
        by = bar_start_y
        bh = int(h * scale)
        
        draw.rectangle(
            [(bx, by), (bx + bar_w, by + bh)],
            fill=(51, 102, 230, 255)  # 主蓝色
        )
    
    return img

def main():
    # 创建输出目录
    output_dir = "Sources/StatBar/Assets/AppIcon.appiconset"
    os.makedirs(output_dir, exist_ok=True)
    
    # 生成各尺寸图标
    for size in SIZES:
        img = create_icon(size)
        output_path = os.path.join(output_dir, f"icon_{size}x{size}.png")
        img.save(output_path, "PNG")
        print(f"Generated: {output_path}")
    
    # 生成 Contents.json
    contents = '''{
  "images" : [
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "16x16",
      "filename" : "icon_16x16.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "16x16",
      "filename" : "icon_32x32.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "32x32",
      "filename" : "icon_32x32.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "32x32",
      "filename" : "icon_64x64.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "128x128",
      "filename" : "icon_128x128.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "128x128",
      "filename" : "icon_256x256.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "256x256",
      "filename" : "icon_256x256.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "256x256",
      "filename" : "icon_512x512.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "512x512",
      "filename" : "icon_512x512.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "512x512",
      "filename" : "icon_1024x1024.png"
    }
  ],
  "info" : {
    "author" : "xcode",
    "version" : 1
  }
}'''
    
    contents_path = os.path.join(output_dir, "Contents.json")
    with open(contents_path, 'w') as f:
        f.write(contents)
    print(f"Generated: {contents_path}")
    
    print("\n✅ All icons generated successfully!")

if __name__ == "__main__":
    main()

File: Scripts/generate-icon.swift
#!/usr/bin/env swift

import Foundation
import CoreGraphics
import ImageIO
import UniformTypeIdentifiers

// 图标尺寸
let sizes: [Int] = [16, 32, 64, 128, 256, 512, 1024]

// 创建图标输出目录
let outputDir = "Sources/StatBar/Assets/AppIcon.appiconset"
try FileManager.default.createDirectory(atPath: outputDir, withIntermediateDirectories: true)

// 为每个尺寸生成图标
for size in sizes {
    let scale = size >= 512 ? 2 : 1
    let actualSize = size >= 512 ? size / 2 : size
    
    let context = CGContext(
        data: nil,
        width: actualSize,
        height: actualSize,
        bitsPerComponent: 8,
        bytesPerRow: 0,
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue
    )!
    
    context.scaleBy(x: CGFloat(actualSize) / 1024, y: CGFloat(actualSize) / 1024)
    
    // 绘制圆角背景
    let bgPath = CGPath(
        roundedRect: CGRect(x: 0, y: 0, width: 1024, height: 1024),
        cornerWidth: 180,
        cornerHeight: 180,
        transform: nil
    )
    context.addPath(bgPath)
    
    // 渐变背景：深蓝到紫色
    let colors: [CGColor] = [
        CGColor(red: 0.2, green: 0.4, blue: 0.9, alpha: 1.0),  // 蓝色
        CGColor(red: 0.5, green: 0.3, blue: 0.9, alpha: 1.0)   // 紫色
    ]
    let gradient = CGGradient(colorsSpace: CGColorSpaceCreateDeviceRGB(), colors: colors as CFArray, locations: [0, 1])!
    context.drawLinearGradient(gradient, start: CGPoint(x: 0, y: 0), end: CGPoint(x: 1024, y: 1024), options: [])
    
    // 绘制 CPU 符号（简化的 CPU 芯片图形）
    context.setFillColor(CGColor(red: 1, green: 1, blue: 1, alpha: 0.95))
    
    // CPU 主体（圆角矩形）
    let cpuBody = CGPath(
        roundedRect: CGRect(x: 312, y: 312, width: 400, height: 400),
        cornerWidth: 60,
        cornerHeight: 60,
        transform: nil
    )
    context.addPath(cpuBody)
    context.fillPath()
    
    // CPU 引脚
    let pinWidth: CGFloat = 40
    let pinHeight: CGFloat = 100
    let pinSpacing: CGFloat = 100
    let pinStartX: CGFloat = 372
    let pinEndX: CGFloat = 652
    
    for i in 0..<3 {
        // 上方引脚
        let topPin = CGRect(x: pinStartX + CGFloat(i) * pinSpacing, y: 212, width: pinWidth, height: pinHeight)
        context.fill(topPin)
        
        // 下方引脚
        let bottomPin = CGRect(x: pinStartX + CGFloat(i) * pinSpacing, y: 712, width: pinWidth, height: pinHeight)
        context.fill(bottomPin)
        
        // 左侧引脚
        context.saveGState()
        context.translateBy(x: 512, y: 512)
        context.rotate(by: .pi / 2)
        context.translateBy(x: -512, y: -512)
        let leftPin = CGRect(x: pinStartX + CGFloat(i) * pinSpacing, y: 212, width: pinWidth, height: pinHeight)
        context.fill(leftPin)
        let rightPin = CGRect(x: pinStartX + CGFloat(i) * pinSpacing, y: 712, width: pinWidth, height: pinHeight)
        context.fill(rightPin)
        context.restoreGState()
    }
    
    // CPU 内部图案（三个竖条表示活动状态）
    context.setFillColor(CGColor(red: 0.2, green: 0.4, blue: 0.9, alpha: 1.0))
    
    let barWidth: CGFloat = 60
    let barSpacing: CGFloat = 90
    let barHeights: [CGFloat] = [140, 200, 120]  // 不同高度表示活动
    let barStartY: CGFloat = 512 - 100
    
    for (index, height) in barHeights.enumerated() {
        let bar = CGRect(
            x: 412 + CGFloat(index) * barSpacing,
            y: barStartY,
            width: barWidth,
            height: height
        )
        context.fill(bar)
    }
    
    // 保存图片
    let image = context.makeImage()!
    let url = URL(fileURLWithPath: "\(outputDir)/icon_\(size)x\(size).png")
    
    let destination = CGImageDestinationCreateWithURL(url as CFURL, UTType.png.identifier as CFString, 1, nil)!
    CGImageDestinationAddImage(destination, image, nil)
    CGImageDestinationFinalize(destination)
    
    print("Generated: icon_\(size)x\(size).png")
}

// 生成 Contents.json
let contentsJson = """
{
  "images" : [
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "16x16",
      "filename" : "icon_16x16.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "16x16",
      "filename" : "icon_32x32.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "32x32",
      "filename" : "icon_32x32.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "32x32",
      "filename" : "icon_64x64.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "128x128",
      "filename" : "icon_128x128.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "128x128",
      "filename" : "icon_256x256.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "256x256",
      "filename" : "icon_256x256.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "256x256",
      "filename" : "icon_512x512.png"
    },
    {
      "idiom" : "mac",
      "scale" : "1x",
      "size" : "512x512",
      "filename" : "icon_512x512.png"
    },
    {
      "idiom" : "mac",
      "scale" : "2x",
      "size" : "512x512",
      "filename" : "icon_1024x1024.png"
    }
  ],
  "info" : {
    "author" : "xcode",
    "version" : 1
  }
}
"""

try contentsJson.write(to: URL(fileURLWithPath: "\(outputDir)/Contents.json"), atomically: true, encoding: .utf8)
print("Generated: Contents.json")
print("\n✅ All icons generated successfully!")
